session.go

```diff
@@ -0,0 +1,233 @@
+package apns
+
+import (
+	"container/list"
+	"errors"
+	"io"
+	"sync"
+)
+
+type SessionError struct {
+	Notification Notification
+	Err          Error
+}
+
+func (s SessionError) Error() string {
+	return s.Err.Error()
+}
+
+type Session interface {
+	Send(n Notification) error
+	Connect() error
+	RequeueableNotifications() []Notification
+	Disconnect()
+	Disconnected() bool
+}
+
+type buffer struct {
+	size int
+	m    sync.Mutex
+	*list.List
+}
+
+func newBuffer(size int) *buffer {
+	return &buffer{size, sync.Mutex{}, list.New()}
+}
+
+func (b *buffer) Add(v interface{}) *list.Element {
+	b.m.Lock()
+	defer b.m.Unlock()
+
+	e := b.PushBack(v)
+
+	if b.Len() > b.size {
+		b.Remove(b.Front())
+	}
+
+	return e
+}
+
+type sessionState int
+
+const (
+	sessionStateNew          sessionState = iota
+	sessionStateConnected    sessionState = iota
```

Comment:
Weird. I tried the above method but it didn't work. Will update.

re: "New" - related to [this comment](https://github.com/timehop/apns/pull/33#discussion_r25801581)